### ISSUE TITLE
Improve model evaluation, reproducibility, and NLP pipeline in SMS Spam Detection project

### ISSUE THREAD

**Issue by Tanishk-Rastogi** [Niketkumardheeryan/ML-CaPsule#1807]
I analyzed the SMS Spam Detection project and found several improvements related to model evaluation correctness, reproducibility, preprocessing quality, and project structure.

### 1. Incorrect model evaluation calls

Several models are initialized correctly but the wrong variable is passed into the `Model()` function.

Example:

```python
gbc = GradientBoostingClassifier()
Model(model, X, y)
```

and:

```python
xgb = XGBClassifier(...)
Model(model, X, y)
```

In these cases, the intended model (`gbc`, `xgb`, etc.) is not actually evaluated. This may lead to incorrect accuracy reporting and misleading model comparisons.

Affected models include:

* AdaBoostClassifier
* GradientBoostingClassifier
* XGBClassifier
* ExtraTreesClassifier

---

### 2. Potential preprocessing/data leakage issues

Some preprocessing and vectorization steps are performed globally before proper train/test splitting, which may affect evaluation reliability.

---

### 3. Missing requirements.txt

The project currently does not provide dependency installation instructions for contributors.

---

### 4. Hardcoded Kaggle dataset path

Dataset loading currently depends on a Kaggle-specific path:

```python
data = pd.read_csv('/kaggle/input/sms-spam-collection-dataset/spam.csv')
```

This affects reproducibility outside Kaggle environments.

---

### 5. Lack of cross-validation and hyperparameter tuning

Most models currently use default parameters and rely on a single train/test split.

---

### 6. README and documentation improvements

README can be improved by adding:

* installation steps
* usage instructions
* dependency list
* project workflow
* model comparison table
* future improvements

---

### 7. Minor code cleanup and spelling inconsistencies

Examples:

* repeated imports
* inconsistent formatting
* spelling issues such as “steeming” instead of “stemming”

---

## Proposed Improvements

* Fix incorrect model evaluation calls
* Refactor preprocessing pipeline using sklearn Pipeline
* Add requirements.txt
* Replace hardcoded dataset paths with relative/local paths
* Add cross-validation support
* Add hyperparameter tuning for major models
* Improve README documentation
* Clean redundant imports and formatting
* Add model persistence using joblib/pickle

---

## Expected Outcome

* Correct and reliable model evaluation
* Improved reproducibility
* Better contributor experience
* Cleaner NLP and ML pipeline
* More maintainable project structure

I would like to work on these improvements under GSSoC.
